### PR TITLE
Bootstrap: detect Windows based on sys.platform

### LINKED ISSUE #73689: Bootstrap fails when Windows has `uname` command
`bootstrap.py` assumes Windows platforms do not have the `uname` command in the PATH. Although almost always true, this doesn't hold if the gnuwin32 tools are installed (e.g. for running LLVM tests), in which case the output is:
```
> uname -s
windows32
> uname -m
i686-pc
```

Bootstrap then gives the following error:
```
> python x.py test src/tools/tidy
Updating only changed submodules
Submodules updated in 0.56 seconds
unknown OS type: windows32
Build completed unsuccessfully in 0:00:00
```

@rustbot modify labels: +A-rustbuild

### PR DESCRIPTION
**Pull request by ajpaverd**
Closes #73689.

### CODE REVIEW

**Comment by retep998** [comment]
What effect does this have on the branches where we detect MSYS/MinGW to build `pc-windows-gnu`?

**Comment by mati865** [comment]
> What effect does this have on the branches where we detect MSYS/MinGW to build `pc-windows-gnu`?

It's going to build `x86_64-pc-windows-msvc` unless the target was was explicitly set. This doesn't seem right.

**Comment by retep998** [comment]
Perhaps we could move the `pc-windows-msvc` fallback to the end of the `ostype` if chain, where if none of the `ostype`s match but `sys.platform` says we are windows then we use `x86-64-pc-windows-msvc`, otherwise we error.

**Comment by ajpaverd** [comment]
> Perhaps we could move the `pc-windows-msvc` fallback to the end of the `ostype` if chain, where if none of the `ostype`s match but `sys.platform` says we are windows then we use `x86-64-pc-windows-msvc`, otherwise we error.

Sorry for the broken patch! Yes, I was just thinking to move this fallback to the end of the `ostype` chain. I'll push an update soon.

File: src/bootstrap/bootstrap.py [hunk 1]
```diff
@@ -236,6 +237,11 @@ def default_build_triple():
         if ostype.endswith('WOW64'):
             cputype = 'x86_64'
         ostype = 'pc-windows-gnu'
+    elif sys.platform == 'win32':
+        # Some Windows platforms might have a `uname` command that returns a
+        # non-standard string (e.g. gnuwin32 tools returns `windows32`). In
+        # these cases, fall back to using sys.platform.
+        return 'x86_64-pc-windows-msvc'
```
Comment (Mark-Simulacrum) [inline src/bootstrap/bootstrap.py:244]: Hm, is there a reason not to directly compare with windows32 here? (FWIW windows32 to me sounds like a 32-bit platform, but maybe that's historical baggage or something?)

Comment (ajpaverd) [reply src/bootstrap/bootstrap.py:244]: I think that's historical baggage of the gnuwin32 tools (they also return `i686-pc` for `uname -m` even on a 64-bit system). I didn't find a 64-bit equivalent, so this is probably not a reliable way to determine the actual architecture.

My thinking here was to provide a default for Windows platforms with any non-standard `uname` value (same as not having `uname`), but I can certainly change this to a direct comparison with windows32. What would be best?

Comment (Mark-Simulacrum) [reply src/bootstrap/bootstrap.py:244]: Hm, so I think this is probably fine -- we can always limit the scope later, if necessary, or add more conditionals.

Comment (ajpaverd) [reply src/bootstrap/bootstrap.py:244]: Thanks for the review!

**Pull request merged**